### PR TITLE
Refine Native Parquet Lib 

### PR DESCRIPTION
**Pull request by QuanWanxx**
1. Fix a bug that native schema uses "Name" property as key to generate Parquet schema.
2. Make Native Parquet lib handle more complex input.
3. Add more tests for Native Parquet Lib.

Refer contained [Native-Parquet](https://github.com/microsoft/FHIR-Analytics-Pipelines/blob/personal/quwan/Refine-Native-Parquet/FhirToDataLake/docs/functional-test/Native-Parquet.md) for detailed information.

### CODE REVIEW

File: FhirToDataLake/native/parquet/csharp/src/ParquetException.cs [hunk 1]
```diff
@@ -28,13 +28,13 @@ private static string GetParquetErrorMessage(int status)
         {
             switch (status)
             {
-                case 10001:
+                case ParquetConverterErrorCodes.ReadInputJsonError:
                     return "Input json is invalid.";
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/csharp/src/ParquetException.cs:32]: define const string for these error message?

File: FhirToDataLake/native/parquet/csharp/src/ParquetException.cs [hunk 2]
```diff
@@ -28,13 +28,13 @@ private static string GetParquetErrorMessage(int status)
...
-                case 10002:
+                case ParquetConverterErrorCodes.WriteToParquetError:
                     return "Write to parquet failed.";
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/csharp/src/ParquetException.cs:34]: failed to do?

File: FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs [hunk 3]
```diff
@@ -46,12 +46,18 @@ public Stream ConvertJsonToParquet(string schemaType, string inputJson)
             // Output buffer pointer
             IntPtr outputPointer = IntPtr.Zero;
 
+            if (inputJson == null)
+            {
+                throw new ParquetException(ParquetConverterErrorCodes.ReadInputJsonError, "Input Json is null.");
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs:51]: use the same string variable as `ParquetException` to keep consistent for the error message.

File: FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs [hunk 4]
```diff
@@ -46,12 +46,18 @@ public Stream ConvertJsonToParquet(string schemaType, string inputJson)
...
+            }
+
             // Get byte counts from input
             int inputSize = Encoding.UTF8.GetByteCount(inputJson);
             StringBuilder errorMessage = new StringBuilder(ErrorMessageSize);
             int status = ConvertJsonToParquet(_nativeConverter, schemaType, inputJson, inputSize, ref outputPointer, out int outputSize, errorMessage);
             if (status != 0)
             {
+                TryReleaseUnmanagedData(ref outputPointer);
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs:60]: do we also need to release memory if the `outputSize` is zero?

File: FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs [hunk 5]
```diff
@@ -46,12 +46,18 @@ public Stream ConvertJsonToParquet(string schemaType, string inputJson)
             // Output buffer pointer
             IntPtr outputPointer = IntPtr.Zero;
 
+            if (inputJson == null)
+            {
+                throw new ParquetException(ParquetConverterErrorCodes.ReadInputJsonError, "Input Json is null.");
+            }
+
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs:53]: Do we need specify error message for Line 90

File: FhirToDataLake/native/parquet/cpp/src/ParquetWriter.cpp [hunk 6]
```diff
@@ -14,7 +14,7 @@ int WriteToParquet(const shared_ptr<arrow::Table> table, byte** outputData, int*
     if (!status.ok())
     {
         string errorDetail = status.ToString();
-        strncpy(errorMessage, errorDetail.c_str(), 200);
+        WriteErrorMessage(errorDetail, errorMessage);
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/cpp/src/ParquetWriter.cpp:17]: why not initilize _readOptions in `ParquetWriter(const unordered_map<string, string>& schemaData)`?

File: FhirToDataLake/native/parquet/cpp/src/ParquetWriter.cpp [hunk 7]
```diff
@@ -49,9 +49,28 @@ int ParquetWriter::RegisterSchema(const string& schemaKey, const string& schemaD
...
 {
+    if (outputData == nullptr)
+    {
+        WriteErrorMessage("Output data pointer is null.", errorMessage);
+        return WriteToParquetError;
+    }
+
+    if (outputLength == nullptr)
+    {
+        WriteErrorMessage("Output data size is null.", errorMessage);
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/cpp/src/ParquetWriter.cpp:60]: Output data size pointer is null?

File: FhirToDataLake/native/parquet/cpp/src/ParquetWriter.cpp [hunk 8]
```diff
@@ -49,9 +49,28 @@ int ParquetWriter::RegisterSchema(const string& schemaKey, const string& schemaD
...
+    if (inputJson == nullptr)
+    {
+        WriteErrorMessage("Input Json data is null.", errorMessage);
+        return ReadInputJsonError;
+    }
+    
     auto schema = _schemaManager.GetSchema(resourceType);
     if (schema == nullptr)
     {
+        WriteErrorMessage("Schema not found for '" + resourceType + "'.", errorMessage);
```
Comment (hongyh13) [inline FhirToDataLake/native/parquet/cpp/src/ParquetWriter.cpp:73]: need check resource type null?

File: FhirToDataLake/native/parquet/cpp/src/SchemaManager.h [hunk 9]
```diff
@@ -16,6 +16,7 @@ const set<string> FhirBooleanTypes { "boolean" };
 shared_ptr<arrow::Field> GenerateStructField(const string& fieldName, const Json::Value& node);
 vector<shared_ptr<arrow::Field>> GenerateSchemaFields(const Json::Value& node);
 bool LoadJson(const string& json, Json::Value* root);
+bool isNullOrWhitespace(const std::string& str);
```
Comment (ACMoretxj) [inline FhirToDataLake/native/parquet/cpp/src/SchemaManager.h:19]: Nit, keep naming convention with existing functions, like `LoadJson` with uppercase first letter.

File: FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp [hunk 10]
```diff
@@ -48,22 +48,37 @@ TEST (SchemaTest, Valid)
...
+    auto brokenSchema = schemaManager.GetSchema(schemaKey);
     EXPECT_TRUE(brokenSchema == nullptr);
-}
\ No newline at end of file
+}
+
+TEST (SchemaTest,  AddAndGetEmptyOrWhiteSpaceSchemaKey)
+{
+    SchemaManager schemaManager;
+    string mockSchema = "{\"Name\": \"Organization\", \"NodePaths\": [\"Organization\"], \"SubNodes\": { \"id\": {\"Name\":\"id\", \"Depth\": 1, \"Type\": \"id\", \"IsLeaf\": true, \"IsRepeated\": false}}, \"Type\": \"Organization\", \"IsRepeated\": false}";
```
Comment (moria97) [inline FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp:78]: Can we also add tests for schema that has some required field missing like id, nodepath, isrepeated, isleaf?

File: FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp [hunk 11]
```diff
@@ -48,22 +48,37 @@ TEST (SchemaTest, Valid)
...
+    int status = schemaManager.AddSchema("", mockSchema);
+    EXPECT_EQ(11001, status);
+
+    status = schemaManager.AddSchema(" ", mockSchema);
+    EXPECT_EQ(11001, status);
+}
```
Comment (moria97) [inline FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp:84]: Maybe also need tests for isnullorwhitespace?

File: FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs (hunk 3 above)
Comment (QuanWanxx) [reply FhirToDataLake/native/parquet/csharp/src/ParquetConverter.cs]: Pre offline discussion, keep current string and move it into rexs.

File: FhirToDataLake/native/parquet/csharp/src/ParquetException.cs (hunk 1 above)
Comment (QuanWanxx) [reply FhirToDataLake/native/parquet/csharp/src/ParquetException.cs]: Updated, thanks.

File: FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp (hunk 11 above)
Comment (QuanWanxx) [reply FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp:84]: Added, thanks.

File: FhirToDataLake/native/parquet/cpp/src/SchemaManager.h (hunk 9 above)
Comment (QuanWanxx) [reply FhirToDataLake/native/parquet/cpp/src/SchemaManager.h]: Fixed, thanks.

File: FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp (hunk 10 above)
Comment (QuanWanxx) [reply FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp:78]: Added more tests in Native CSharp project, here is no "required" field, the field can be missed as long as it matched with input data.

The default property type is string, default IsRepeated and IsLeaf is false. 

I added a simplest schema test file, only contains all used sections to correctly convert a data.

File: FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp [hunk 12]
```diff
@@ -48,22 +48,49 @@ TEST (SchemaTest, Valid)
...
-    auto emptySchema = schemaManager.GetSchema("Empty");
+    int status = schemaManager.AddSchema(schemaKey, empty);
     EXPECT_EQ(11001, status);
+
+    auto emptySchema = schemaManager.GetSchema(schemaKey);
     EXPECT_TRUE(emptySchema == nullptr);
 }
 
-TEST (SchemaTest, Invalid)
+TEST (SchemaTest,  AddAndGetNoneJsonSchemaContent)
```
Comment (moria97) [inline FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp:63]: nit, double spaces here

Comment (moria97) [reply FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp:63]: same for following two functions

File: FhirToDataLake/docs/functional-test/Native-Parquet.md [hunk 13]
```diff
@@ -0,0 +1,95 @@
...
+2. SchemaManager
+    - int AddSchema(const string& schemaKey, const string& schemaJson);
+      1. Correct add schema with given schema key.
+      2. **[Add]** Can return false if schema content is empty, white space or invalid Json.
+      3. **[Add]** Can return false if schema key is empty or white space.
+
+    - shared_ptr<arrow::Schema> GetSchema(const string& schemaKey);
+      1. Correct return schema with given schema key.
+
+## CSharp
```
Comment (moria97) [inline FhirToDataLake/docs/functional-test/Native-Parquet.md:71]: Can we add a work item to track a e2e test if FHIR server has invalid data to transform? like a field with wrong type

**Review by moria97 — APPROVED** [review]

**Review by hongyh13 — APPROVED** [review]

File: FhirToDataLake/docs/functional-test/Native-Parquet.md (hunk 13 above)
Comment (QuanWanxx) [reply FhirToDataLake/docs/functional-test/Native-Parquet.md:71]: Sure, create this [item](https://microsofthealth.visualstudio.com/Health/_workitems/edit/97856)

File: FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp (hunk 12 above)
Comment (QuanWanxx) [reply FhirToDataLake/native/parquet/cpp/test/SchemaManagerTests.cpp]: Fixed, thanks.

**Pull request merged**